File: src/test/java/no/nav/fo/veilarbregistrering/registrering/bruker/BrukerTilstandServiceTest.java
package no.nav.fo.veilarbregistrering.registrering.bruker;

import no.nav.common.featuretoggle.UnleashService;
import no.nav.fo.veilarbregistrering.arbeidssoker.Formidlingsgruppe;
import no.nav.fo.veilarbregistrering.oppfolging.OppfolgingGateway;
import no.nav.fo.veilarbregistrering.oppfolging.Oppfolgingsstatus;
import no.nav.fo.veilarbregistrering.oppfolging.Rettighetsgruppe;
import no.nav.fo.veilarbregistrering.oppfolging.Servicegruppe;
import no.nav.fo.veilarbregistrering.sykemelding.SykemeldingService;
import no.nav.fo.veilarbregistrering.sykemelding.SykmeldtInfoData;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class BrukerTilstandServiceTest {

    private OppfolgingGateway oppfolgingGateway;
    private SykemeldingService sykemeldingService;
    private UnleashService unleashService;

    private BrukerTilstandService brukerTilstandService;

    @BeforeEach
    public void setUp() {
        oppfolgingGateway = mock(OppfolgingGateway.class);
        sykemeldingService = mock(SykemeldingService.class);
        unleashService = mock(UnleashService.class);

        brukerTilstandService = new BrukerTilstandService(oppfolgingGateway, sykemeldingService, unleashService);
    }

    @Test
    public void brukersTilstand_skal_gi_sperret_når_toggle_ikke_er_skrudd_på() {
        Oppfolgingsstatus oppfolgingsstatus = new Oppfolgingsstatus(
                false,
                false,
                true,
                Formidlingsgruppe.of("IARBS"),
                Servicegruppe.of("VURDI"),
                Rettighetsgruppe.of("IYT"));
        when(oppfolgingGateway.hentOppfolgingsstatus(any())).thenReturn(oppfolgingsstatus);

        SykmeldtInfoData sykeforlop = new SykmeldtInfoData(null, false);
        when(sykemeldingService.hentSykmeldtInfoData(any())).thenReturn(sykeforlop);

        when(unleashService.isEnabled(any())).thenReturn(false);

        BrukersTilstand brukersTilstand = brukerTilstandService.hentBrukersTilstand(any());

        assertThat(brukersTilstand.getRegistreringstype()).isEqualTo(RegistreringType.SPERRET);
    }

    @Test
    public void brukersTilstand_skal_gi_sykmeldtRegistrering_når_toggle_er_enablet() {
        Oppfolgingsstatus oppfolgingsstatus = new Oppfolgingsstatus(
                false,
                false,
                true,
                Formidlingsgruppe.of("IARBS"),
                Servicegruppe.of("VURDI"),
                Rettighetsgruppe.of("IYT"));
        when(oppfolgingGateway.hentOppfolgingsstatus(any())).thenReturn(oppfolgingsstatus);

        SykmeldtInfoData sykeforlop = new SykmeldtInfoData(null, false);
        when(sykemeldingService.hentSykmeldtInfoData(any())).thenReturn(sykeforlop);

        when(unleashService.isEnabled(any())).thenReturn(true);

        BrukersTilstand brukersTilstand = brukerTilstandService.hentBrukersTilstand(any());

        assertThat(brukersTilstand.getRegistreringstype()).isEqualTo(RegistreringType.SYKMELDT_REGISTRERING);
    }

    @Test
    public void brukersTilstand_hvor_med_og_uten_maksdato_gir_ulike_svar() {
        Oppfolgingsstatus oppfolgingsstatus = new Oppfolgingsstatus(
                false,
                false,
                true,
                Formidlingsgruppe.of("IARBS"),
                Servicegruppe.of("VURDI"),
                Rettighetsgruppe.of("IYT"));
        when(oppfolgingGateway.hentOppfolgingsstatus(any())).thenReturn(oppfolgingsstatus);

        SykmeldtInfoData sykeforlop = new SykmeldtInfoData(null, false);
        when(sykemeldingService.hentSykmeldtInfoData(any())).thenReturn(sykeforlop);

        when(unleashService.isEnabled(any())).thenReturn(true);

        BrukersTilstand brukersTilstand = brukerTilstandService.hentBrukersTilstand(any(), true);

        assertThat(brukersTilstand.getRegistreringstype()).isEqualTo(RegistreringType.SYKMELDT_REGISTRERING);
    }

    @Test
    public void brukersTilstand_hvor_med_og_uten_maksdato_gir_like_svar() {
        Oppfolgingsstatus oppfolgingsstatus = new Oppfolgingsstatus(
                false,
                false,
                true,
                Formidlingsgruppe.of("IARBS"),
                Servicegruppe.of("VURDI"),
                Rettighetsgruppe.of("IYT"));
        when(oppfolgingGateway.hentOppfolgingsstatus(any())).thenReturn(oppfolgingsstatus);

        SykmeldtInfoData sykeforlop = new SykmeldtInfoData(LocalDate.now().minusWeeks(10).toString(), true);
        when(sykemeldingService.hentSykmeldtInfoData(any())).thenReturn(sykeforlop);

        when(unleashService.isEnabled(any())).thenReturn(true);

        BrukersTilstand brukersTilstand = brukerTilstandService.hentBrukersTilstand(any(), true);

        assertThat(brukersTilstand.getRegistreringstype()).isEqualTo(RegistreringType.SYKMELDT_REGISTRERING);
    }
}

File: src/test/java/no/nav/fo/veilarbregistrering/registrering/bruker/BrukerTilstandServiceTest.kt
package no.nav.fo.veilarbregistrering.registrering.bruker

import io.mockk.every
import io.mockk.mockk
import no.nav.fo.veilarbregistrering.arbeidssoker.Formidlingsgruppe
import no.nav.fo.veilarbregistrering.bruker.AktorId
import no.nav.fo.veilarbregistrering.bruker.Bruker
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.oppfolging.OppfolgingGateway
import no.nav.fo.veilarbregistrering.oppfolging.Oppfolgingsstatus
import no.nav.fo.veilarbregistrering.oppfolging.Rettighetsgruppe
import no.nav.fo.veilarbregistrering.oppfolging.Servicegruppe
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class BrukerTilstandServiceTest {
    private lateinit var oppfolgingGateway: OppfolgingGateway
    private lateinit var brukerRegistreringRepository: BrukerRegistreringRepository
    private lateinit var brukerTilstandService: BrukerTilstandService
    @BeforeEach
    fun setUp() {
        oppfolgingGateway = mockk()
        brukerRegistreringRepository = mockk(relaxed = true)
        brukerTilstandService = BrukerTilstandService(
            oppfolgingGateway,
            brukerRegistreringRepository
        )
    }

    @Test
    fun `brukersTilstand skal gi sykmeldtRegistrering når bruker er sykmeldt med arbeidsgiver`() {
        val oppfolgingsstatus = Oppfolgingsstatus(
            false,
            false,
            true,
            Formidlingsgruppe.of("IARBS"),
            Servicegruppe.of("VURDI"),
            Rettighetsgruppe.of("IYT")
        )
        every { oppfolgingGateway.hentOppfolgingsstatus(any()) } returns oppfolgingsstatus
        val brukersTilstand = brukerTilstandService.hentBrukersTilstand(testBruker)
        Assertions.assertThat(brukersTilstand.registreringstype).isEqualTo(RegistreringType.SYKMELDT_REGISTRERING)
    }

    companion object {
        private val testBruker = Bruker.of(Foedselsnummer.of("11019141466"), AktorId("1"), emptyList())
    }
}